File: export/client_configuration.cs

using MessagePack;
using System.Collections.Generic;
/// <summary> 配置 </summary>
namespace Configuration 
{
    /// <summary>
    /// 测试
    /// </summary>*/
    [MessagePackObject]
    public class Test 
    {
        /// <summary> id </summary>
        [Key(0)]
        public Int64? id;
        /// <summary> 数值 </summary>
        [Key(1)]
        public Int64? num;
        /// <summary> 字符串 </summary>
        [Key(2)]
        public string? str;
    }

    /// <summary>
    /// 属性
    /// </summary>*/
    [MessagePackObject]
    public class Attr 
    {
        /// <summary> 属性ID </summary>
        [Key(0)]
        public Int64? attrId;
        /// <summary> 属性值 </summary>
        [Key(1)]
        public Int64? value;
    }

    /// <summary>
    /// 三维坐标
    /// </summary>*/
    [MessagePackObject]
    public class Vector3 
    {
        /// <summary> 坐标X </summary>
        [Key(0)]
        public float? x;
        /// <summary> 坐标Y </summary>
        [Key(1)]
        public float? y;
        /// <summary> 坐标Z </summary>
        [Key(2)]
        public float? z;
    }

    [MessagePackObject]
    public class Monster 
    {
        /// <summary> 怪物Id </summary>
        [Key(0)]
        public Int64? monsterId;
        /// <summary> 怪物名 </summary>
        [Key(1)]
        public string? name;
        /// <summary> 怪物等级 </summary>
        [Key(2)]
        public Int64? level;
        /// <summary> 技能列表 id#id </summary>
        [Key(3)]
        public Int64[]? skills;
        /// <summary> 怪物模型 </summary>
        [Key(4)]
        public string? model;
        /// <summary> 坐标 </summary>
        [Key(5)]
        public Vector3? pos;
        /// <summary> 头像 </summary>
        [Key(6)]
        public string? head;
    }
    public class TypeNames {
        /// <summary> G-怪物配置 </summary>
        public static string monster = "monster";
    }


    public class Struct {
        /// <summary> G-怪物配置 </summary>
        public Dictionary<string, Monster>? monster;

    };

}

File: export/ClientConfiguration.cpp

#include <tuple>
#include <string>
#include <vector>
#include <optional>
#include <unordered_map>

#include "ClientConfiguration.h"

namespace Gen
{
    /* 配置 */
    namespace Configuration
    {
        using namespace std;
#ifdef CLIENT
        std::optional<Test::Tuple> TestEncode(std::optional<Test> &obj)
        {
            if (obj.has_value())
            {

                auto &oValue = obj.value();
                return Test::Tuple(oValue.num32, oValue.uNum32, oValue.id64, oValue.uId64, oValue.str);
            }
            else
            {
                return std::nullopt;
            }
        }
        std::optional<Test> TestDecode(std::optional<Test::Tuple> &t)
        {
            std::optional<Test> obj;
            if (t.has_value())
            {
                obj = Test();
                auto &oValue = obj.value();
                auto &tValue = t.value();
                oValue.num32 = std::get<0>(tValue);
                oValue.uNum32 = std::get<1>(tValue);
                oValue.id64 = std::get<2>(tValue);
                oValue.uId64 = std::get<3>(tValue);
                oValue.str = std::get<4>(tValue);
            }
            return obj;
        }

        std::optional<Attr::Tuple> AttrEncode(std::optional<Attr> &obj)
        {
            if (obj.has_value())
            {

                auto &oValue = obj.value();
                return Attr::Tuple(oValue.attrId, oValue.value);
            }
            else
            {
                return std::nullopt;
            }
        }
        std::optional<Attr> AttrDecode(std::optional<Attr::Tuple> &t)
        {
            std::optional<Attr> obj;
            if (t.has_value())
            {
                obj = Attr();
                auto &oValue = obj.value();
                auto &tValue = t.value();
                oValue.attrId = std::get<0>(tValue);
                oValue.value = std::get<1>(tValue);
            }
            return obj;
        }

        std::optional<Vector3::Tuple> Vector3Encode(std::optional<Vector3> &obj)
        {
            if (obj.has_value())
            {

                auto &oValue = obj.value();
                return Vector3::Tuple(oValue.x, oValue.y, oValue.z);
            }
            else
            {
                return std::nullopt;
            }
        }
        std::optional<Vector3> Vector3Decode(std::optional<Vector3::Tuple> &t)
        {
            std::optional<Vector3> obj;
            if (t.has_value())
            {
                obj = Vector3();
                auto &oValue = obj.value();
                auto &tValue = t.value();
                oValue.x = std::get<0>(tValue);
                oValue.y = std::get<1>(tValue);
                oValue.z = std::get<2>(tValue);
            }
            return obj;
        }

        std::optional<Monster::Tuple> MonsterEncode(std::optional<Monster> &obj)
        {
            if (obj.has_value())
            {

                auto &oValue = obj.value();
                std::optional<std::vector<std::optional<int64>>> skillsArr;
                if (oValue.skills.has_value())
                {
                    skillsArr = std::vector<std::optional<int64>>();
                    auto &value = skillsArr.value();
                    auto &skills = oValue.skills.value();
                    for (auto i = 0; i < skills.size(); ++i)
                    {
                        value.push_back(skills[i]);
                    }
                }
                return Monster::Tuple(oValue.monsterId, oValue.name, oValue.level, skillsArr, oValue.model, Vector3Encode(oValue.pos), oValue.head);
            }
            else
            {
                return std::nullopt;
            }
        }
        std::optional<Monster> MonsterDecode(std::optional<Monster::Tuple> &t)
        {
            std::optional<Monster> obj;
            if (t.has_value())
            {
                obj = Monster();
                auto &oValue = obj.value();
                auto &tValue = t.value();
                auto &skills_t = std::get<3>(tValue);
                if (skills_t.has_value())
                {
                    oValue.skills = std::vector<std::optional<int64>>();
                    auto &skills = skills_t.value();
                    auto &value = oValue.skills.value();
                    for (auto i = 0; i < skills.size(); ++i)
                    {
                        value.push_back(skills[i]);
                    }
                }
                oValue.monsterId = std::get<0>(tValue);
                oValue.name = std::get<1>(tValue);
                oValue.level = std::get<2>(tValue);
                oValue.model = std::get<4>(tValue);
                oValue.pos = Vector3Decode(std::get<5>(tValue));
                oValue.head = std::get<6>(tValue);
            }
            return obj;
        }

#endif // CLIENT    }
}